cardano-api/src/Cardano/Api/ProtocolParameters.hs

```diff
@@ -212,14 +226,380 @@ data ProtocolParameters =
+      ShelleyBasedEraMary ->
+         parameterExists minUtxo (ProtocolParamsErrMustDefineMinUTxo anyEra)
+  --  ShelleyBasedEraAlonzo -> do
+  --     parameterExists utxoCostPerByte (ProtocolParamsErrMustDefineCostPerByte anyEra)
+  --     parameterExists costModels (ProtocolParamsErrMustDefineCostModel anyEra)
+  --     parameterExists prices (ProtocolParamsErrMustDefineScriptExecPrices anyEra)
+  --     parameterExists maxTxExUnits (ProtocolParamsErrMustDefineMaxTxExec anyEra)
+  --     parameterExists maxBlockExUnits (ProtocolParamsErrMustDefineMaxBlockExec anyEra)
+  --     parameterExists maxValueSize (ProtocolParamsErrMustDefineMaxValueSize anyEra)
+
+  -- | Check if a protocol parameter was specified and fail if not.
+  parameterExists
+    :: Maybe a
+    -> ProtocolParametersError
+    -> Either ProtocolParametersError ()
+  parameterExists value err =
+    maybe (Left err) Right value >> Right ()
+
+
+newtype MaxTxExecutionUnits =
+    MaxTxExecutionUnits { unMaxTxExecutionUnits :: ExecutionUnits}
+    deriving (Eq, Show)
+
+_fromMaxTxExec :: Alonzo.ExUnits -> MaxTxExecutionUnits
+_fromMaxTxExec (Alonzo.ExUnits mMem mSteps) =
+  MaxTxExecutionUnits $ ExecutionUnits mMem mSteps
+
+_toTxExecUnits :: MaxTxExecutionUnits -> Alonzo.ExUnits
+_toTxExecUnits (MaxTxExecutionUnits (ExecutionUnits mMem mSteps)) =
+  Alonzo.ExUnits mMem mSteps
+
+instance ToJSON MaxTxExecutionUnits where
+  toJSON (MaxTxExecutionUnits (ExecutionUnits space time)) =
+    object [ "maxTxExecutionUnits" .=
+                object ["space" .= space, "time" .= time]
+           ]
+
+instance FromJSON MaxTxExecutionUnits where
+  parseJSON = withObject "MaxTxExecutionUnits" $ \o -> do
+    obj <- o .: "maxTxExecutionUnits"
+    MaxTxExecutionUnits
+      <$> (ExecutionUnits <$> obj .: "space" <*> obj .: "time")
+
+newtype MaxBlockExecutionUnits =
+    MaxBlockExecutionUnits { unMaxBlockExecutionUnits :: ExecutionUnits}
+    deriving (Eq, Show)
+
+_fromMaxBlockExec :: Alonzo.ExUnits -> MaxBlockExecutionUnits
+_fromMaxBlockExec (Alonzo.ExUnits mMem mSteps) =
+  MaxBlockExecutionUnits $ ExecutionUnits mMem mSteps
+
+_toExUnits :: MaxBlockExecutionUnits -> Alonzo.ExUnits
+_toExUnits (MaxBlockExecutionUnits (ExecutionUnits mMem mSteps)) =
+  Alonzo.ExUnits mMem mSteps
+
+instance ToJSON MaxBlockExecutionUnits where
+  toJSON (MaxBlockExecutionUnits (ExecutionUnits space time)) =
+    object [ "maxBlockExecutionUnits" .=
+                object ["space" .= space, "time" .= time]
+           ]
+
+instance FromJSON MaxBlockExecutionUnits where
+  parseJSON = withObject "MaxBlockExecutionUnits" $ \o -> do
+    obj <- o .: "maxBlockExecutionUnits"
+    MaxBlockExecutionUnits
+      <$> (ExecutionUnits <$> obj .: "space" <*> obj .: "time")
+
+data ExecutionUnits
+    = ExecutionUnits { space :: Word64
+                     , time :: Word64
+                     } deriving (Eq, Show)
+
+
+newtype CostModel = CostModel (Map AnyScriptLanguage Cost)
```

Comment:
So I think this name should be for an individual cost model rather than the collection of them all.

If `protocolParamCostModels` is adjusted as I suggest to the the collection then it's ok if `CostModel` is a single one.

The cost model representation will be opaque, wrapping the type from the lib, with just serialisation instances.